pkgs/swarmauri/swarmauri/llms/concrete/OpenAIModel.py

```diff
@@ -106,3 +117,134 @@ def predict(
+        result = json.loads(response.model_dump_json())
+        message_content = result["choices"][0]["message"]["content"]
+        conversation.add_message(AgentMessage(content=message_content))
+
+        return conversation
+
+    def stream(
+        self,
+        conversation,
+        temperature=0.7,
+        max_tokens=256,
+        stop: List[str] = [],
+    ) -> Iterator[str]:
+        """Synchronously stream the response token by token"""
+        formatted_messages = self._format_messages(conversation.history)
+
+        stream = self.client.chat.completions.create(
+            model=self.name,
+            messages=formatted_messages,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            stream=True,
+            stop=stop,
+        )
+
+        collected_content = []
+        for chunk in stream:
+            if chunk.choices[0].delta.content is not None:
+                content = chunk.choices[0].delta.content
+                collected_content.append(content)
+                yield content
+
+        full_content = "".join(collected_content)
+        conversation.add_message(AgentMessage(content=full_content))
+
+    async def astream(
+        self,
+        conversation,
+        temperature=0.7,
+        max_tokens=256,
+        stop: List[str] = [],
+    ) -> AsyncIterator[str]:
+        """Asynchronously stream the response token by token"""
+        formatted_messages = self._format_messages(conversation.history)
+
+        stream = await self.async_client.chat.completions.create(
+            model=self.name,
+            messages=formatted_messages,
+            temperature=temperature,
+            max_tokens=max_tokens,
+            stream=True,
+            stop=stop,
+        )
+
+        collected_content = []
+        async for chunk in stream:
+            if chunk.choices[0].delta.content is not None:
+                content = chunk.choices[0].delta.content
+                collected_content.append(content)
+                yield content
+
+        full_content = "".join(collected_content)
+        conversation.add_message(AgentMessage(content=full_content))
+
+    def batch(
+        self,
+        conversations: List,
+        temperature=0.7,
+        max_tokens=256,
+        enable_json=False,
+        stop: List[str] = [],
+    ) -> List:
+        """Synchronously process multiple conversations"""
+        return [
+            self.predict(
+                conv,
+                temperature=temperature,
+                max_tokens=max_tokens,
+                enable_json=enable_json,
+                stop=stop,
+            )
+            for conv in conversations
+        ]
+
+    async def abatch(
+        self,
+        conversations: List,
+        temperature=0.7,
+        max_tokens=256,
+        enable_json=False,
+        stop: List[str] = [],
+    ) -> List:
+        """Process multiple conversations in parallel"""
```

Comment:
use a semaphore here to better control how many processes are able to run concurrently